[Builtins] Added lists and tuples

effectfully opened the pull request:
Don't look here yet.

effectfully reviewed (commented):
(still don't look here)

effectfully commented on plutus-core/untyped-plutus-core/test/Evaluation/Machines/Budget/Examples/InterList/FoldrInterList.plc.golden:
```diff
@@ -3,9 +3,9 @@
 | mem: 0
 })
 | BAST causes ({ cpu: 0
-| mem: 46
+| mem: 116
```
And in other places. Really don't get it.

effectfully replied:
I've figured it out: we now have a general instance `ExMemoryUsage` instance for `[a]` and we define the `ExMemoryUsage Text` in terms of `String`, which has changed and we use `Text` in the names of variables, hence the change.
@kwxm do we care?

kwxm replied:
> @kwxm do we care?

I don't think so.  We may have to have a charge for the overall size of the AST, but it doesn't particularly matter if the size changes a bit (and maybe we could base such a charge on the serialised size, instead of doing a potentially expensive traversal of the deserialised AST).  The stuff we're using on the chain shouldn't contain any variable names anyway, and we're probably not going to be using strings much.  I've got rid of `BAST` in my recent PR and replaced it with something called `BStartup` which is supposed to account for startup costs, but I don't have enough data to work out what that should be in general and it's currently just a constant (but that may change later).

effectfully replied:
Thanks, I also thought it probably wouldn't matter.

effectfully commented on plutus-core/untyped-plutus-core/test/Evaluation/Machines/Budget/Examples/InterList/InterNil.plc.golden:
```diff
@@ -3,9 +3,9 @@
 | mem: 0
 })
 | BAST causes ({ cpu: 0
-| mem: 3
+| mem: 9
```
This one doesn't even use any built-in type.

effectfully commented on plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs:
```diff
@@ -187,3 +198,44 @@ instance (GShow uni, GEq uni, uni `Includes` Integer) => ToBuiltinMeaning uni Ex
+        sndPlc (SomeValueArg _ (SomeValueArg uniB (SomeValueRes _ (_, y)))) =
+            Opaque . fromConstant . Some $ ValueOf uniB y
+    toBuiltinMeaning Swap = makeBuiltinMeaning swapPlc mempty where
+        swapPlc :: SomeValueN uni (,) '[a, b] -> SomeValueN uni (,) '[b, a]
```
Made inference work with polymorphic built-in types just like it works with regular type variables!

effectfully commented on plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs:
```diff
@@ -187,3 +198,44 @@ instance (GShow uni, GEq uni, uni `Includes` Integer) => ToBuiltinMeaning uni Ex
+        swapPlc (SomeValueArg uniA (SomeValueArg uniB (SomeValueRes _ (x, y)))) =
+            SomeValueArg uniB (SomeValueArg uniA (SomeValueRes (DefaultUniTuple uniB uniA) (y, x)))
+    toBuiltinMeaning SwapEls = makeBuiltinMeaning swapElsPlc mempty where
+        -- swapElsPlc : [([a], Bool)] -> [(Bool, [a])]
+        swapElsPlc
+            :: a ~ Opaque term (TyVarRep ('TyNameRep "a" 0))
+            => SomeValueN uni [] '[SomeValueN uni (,) '[SomeValueN uni [] '[a], Bool]]
+            -> SomeValueN uni [] '[SomeValueN uni (,) '[Bool, SomeValueN uni [] '[a]]]
```
![We need to go deeper](https://wiki.rtzra.ru/_media/qigong/we_need_to_go_deeper.jpg)

effectfully commented on plutus-core/plutus-core/examples/PlutusCore/Examples/Data/TreeForest.hs:
```diff
@@ -218,7 +219,7 @@ forestNil = runQuote $ normalizeTypesIn =<< do
 -- > /\(a :: *) -> \(tr : tree a) (fr : forest a)
 -- >     wrapForest [a] /\(r :: *) -> \(z : r) (f : tree a -> forest a -> r) -> f tr fr
-forestCons :: Term TyName Name uni fun ()
+forestCons :: HasUniApply uni => Term TyName Name uni fun ()
```
Type normalization now requires that.

kwxm commented:
I made you a "Don't look here yet" label since that appears in your PRs quite often.  I'm wondering if we should have a "Now it's OK to look" one as well...

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Constant/Typed.hs:
```diff
@@ -329,12 +326,14 @@ class KnownTypeAst uni (a :: k) where
 -- See Note [KnownType's defaults].
 -- | A default implementation of 'toTypeAst' for built-in types.
-toBuiltinTypeAst :: uni `Includes` a => proxy a -> Type TyName uni ()
+toBuiltinTypeAst :: uni `Contains` a => proxy a -> Type TyName uni ()
```
Note to self: that was a stupid idea.

effectfully replied:
Note to self: it was a stupid idea that the original idea was a stupid idea, 'cause it in fact made sense and `toBuiltinTypeAst` is kind-polymorphic now.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:
```diff
@@ -46,7 +53,7 @@ data Kind ann
 
 -- | A 'Type' assigned to expressions.
-data Type tyname uni ann
+data Type tyname (uni :: GHC.Type -> GHC.Type) ann
```
We need to be explicit here now because of enabled `PolyKinds`.

effectfully commented on plutus-tx/src/PlutusTx/Lift/Instances.hs:
```diff
@@ -52,7 +53,7 @@ instance Typeable uni (->) where
 -- Primitives
 
 typeRepBuiltin
-    :: forall a uni fun. uni `PLC.Includes` a
+    :: forall (a :: GHC.Type) uni fun. uni `PLC.Includes` a
     => Proxy a -> RTCompile uni fun (Type TyName uni ())
 typeRepBuiltin (_ :: Proxy a) = pure $ mkTyBuiltin @a ()
```
`mkTyBuiltin` is kind-polymorphic. Should this one be too? Or perhaps we don't need it here? Pinging @michaelpj

effectfully commented:
I've updated all dependencies apart from the NEAT tests in the metatheory.

The way I updated the NEAT tests in Plutus Core is by adding `TyListG` to `TypeBuiltinG`, which requires only a few changes to the existing machinery. We can't just keep `TyStringG`, because during type normalization that normalizes to something else (as per `Note [Normalization of built-in types]`) and so we have to actually allow embedding the list type directly into `TypeG` via `TyBuiltinG`.

However that just destroys the metatheory, 'cause it does not support built-in types of kinds other than `*` (which is reasonable, we didn't have them previously). What's worse is that errors that I'm getting a very unspecific about what and where went wrong:

```
NEAT
  type-level: FAIL (0.01s)
    generators/PlutusCore/Generators/NEAT/Spec.hs:446:
    Counterexample found: TyForallG (Type ()) (TyFunG (TyVarG FZ) (TyFunG (TyVarG FZ) (TyFunG (TyVarG FZ) (TyAppG (TyBuiltinG TyListG) (TyVarG FZ) (Type ()))))) :: (type)
  term-level: Constant () (Some (ValueOf unit ()))
(con ())
Error ()
error
Error ()
error
Error ()
error
Error ()
error
Error ()
error
Error ()
error
Error ()
error
Error ()
error
Error ()
error
FAIL (2.91s)
    generators/PlutusCore/Generators/NEAT/Spec.hs:446:
    Counterexample found: ErrorG (TyAppG (TyLamG (TyVarG FZ)) (TyAppG (TyLamG (TyVarG FZ)) (TyAppG (TyLamG (TyBuiltinG TyUnitG)) (TyBuiltinG TyListG) (KindArrow () (Type ()) (Type ()))) (Type ())) (Type ())) :: TyBuiltinG TyUnitG
```

What's even worse is that I haven't added calls to `error` or any unsafety in general to any function related to Plutus Core's type level in the metatheory. I.e. neither Haskell's nor Agda's type checker has warned me about the fact that the contents of the set of built-in types has changed and they now need to be handled differently. How so? Well, here's how we define raw types in Haskell (we're interested only in built-in types, hence other constructors are omitted):

```haskell
data RType =
          <...>
           | RTyCon (SomeTypeIn DefaultUni)
```

And here's how we define them in Agda:

```agda
data RawTy : Set where
  <...>
  con : TyCon → RawTy
```

and then we connect the two (right?) via the following declaration:

```
{-# COMPILE GHC RawTy = data RType (RTyVar | RTyFun | RTyPi | RTyLambda | RTyApp | RTyCon | RTyMu) #-}
```

I.e. we implicitly convert (Agda does a lot of `unsafeCoerce`) the contents of the Haskell's `RTyCon` constructor to the one of the Agda's `con`. But they are wildly different! Here's what we have in Agda:

```agda
data TyCon : Set where
  integer    : TyCon
  bytestring : TyCon
  string     : TyCon
  char       : TyCon
  unit       : TyCon
  bool       : TyCon
```

Previously that worked because of the clever pattern synonyms:

```agda
{-# FOREIGN GHC {-# LANGUAGE GADTs, PatternSynonyms #-}                #-}
{-# FOREIGN GHC import PlutusCore                                      #-}
{-# FOREIGN GHC type TypeBuiltin = SomeTypeIn DefaultUni               #-}
{-# FOREIGN GHC pattern TyInteger    = SomeTypeIn DefaultUniInteger    #-}
{-# FOREIGN GHC pattern TyByteString = SomeTypeIn DefaultUniByteString #-}
{-# FOREIGN GHC pattern TyString     = SomeTypeIn DefaultUniString     #-}
{-# FOREIGN GHC pattern TyChar       = SomeTypeIn DefaultUniChar       #-}
{-# FOREIGN GHC pattern TyUnit       = SomeTypeIn DefaultUniUnit       #-}
{-# FOREIGN GHC pattern TyBool       = SomeTypeIn DefaultUniBool       #-}
{-# COMPILE GHC TyCon = data TypeBuiltin (TyInteger | TyByteString | TyString | TyChar | TyUnit | TyBool) #-}
```

But now that we actually generate a built-in type of kind `* -> *` (the type of lists), no way we can convert that to anything in the metatheory without updating the metatheory. Which does look like a huge task.

And in future we should tweak `metatheory` to either support extensible universes directly or to have explicit conversion between extensible universes and the non-extensible one (either on the side of Haskell or Agda -- does not matter). So that we don't do this conversion implicitly (we'll still convert something implicitly, but at least converting from one representation to the same one in a different language is a much safer thing than crossing the language barrier and changing the representation at the same time).

The best solution I can think of is to filter out generated `TypeG` whenever they have `TyListG` in them **only in the metatheory's NEAT tests** ('cause in Plutus Core that is completely fine). Just for now before we support built-in types of kinds other than `*` in the metatheory.

But I haven't looked into how to do that. @jmchapman could you help me with that please? Ideally by updating the metatheory-s NEAT tests directly, but even just providing some suggestions would be helpful.

jmchapman commented:
filtering out types containing lists in metatheory tests:

```
bigTest :: (Check t a, Enumerable a)
        => String -> GenOptions -> t -> (t -> a -> Assertion) -> TestTree
bigTest s GenOptions{..} t f = testCaseInfo s $ do
  as <- search' genMode genDepth (\a ->  check t a)
  _  <- traverse (f t) as
  return $ show (length as)
```
both `plutus-core` and `plutus-metatheory` tests both use this function which filters out stuff that doesn't kind check or type check using `check`. How about making a copy of this function for metatheory use only that also checks if the type (which annoyingly could be the `t` or the `a`) contains a list?

jmchapman commented:
unhelpful error messages: I wonder if my catching of errors and turning them into `error` is making this worse...

```
plutus-core/generators/PlutusCore/Generators/NEAT/Spec.hs

-- handle a user error and turn it back into an error term
handleError :: Type TyName DefaultUni ()
       -> U.ErrorWithCause (U.EvaluationError user internal) term
       -> Either (U.ErrorWithCause (U.EvaluationError user internal) term)
                 (Term TyName Name DefaultUni DefaultFun ())
handleError ty e = case U._ewcError e of
  U.UserEvaluationError     _ -> return (Error () ty)
  U.InternalEvaluationError _ -> throwError e
```

effectfully commented:
> both plutus-core and plutus-metatheory tests use this function which filters out stuff that doesn't kind check or type check using check. How about making a copy of this function for metatheory use only that also checks if the type (which annoyingly could be the t or the a) contains a list?

Sounds like something that should work, thanks. I'll look into it once I'm done with docs, but feel free to do that yourself.

> Or maybe this is the wrong bit of code, I think the agda model may turn any error into an error term.

It looks irrelevant, 'cause surely NEAT tests should fail before we get into testing evaluation? At type normalization and type checking phases. Although if something that does not kind/type check gets filtered out... Dunno.

jmchapman commented:
> Sounds like something that should work, thanks. I'll look into it once I'm done with docs, but feel free to do that yourself.

I'll give it a shot.

jmchapman commented:
note: I didn't push the agda2hs generated haskell files.

effectfully commented:
Thank you!

> note: I didn't push the agda2hs generated haskell files.

I have them committed already, so I'll do it myself.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Constant/Typed.hs:
```diff
@@ -284,60 +385,28 @@ type HasConstant term = (AsConstant term, FromConstant term)
-but then we need to recover the original handy definitions and make the API and
-the code bloated (instances are more verbose with @DerivingVia@).
-
-We don't use @default@ in 'KnownTypeAst', because an attempt to write
-
-    default toTypeAst :: (k ~ GHC.Type, uni `Includes` a) => proxy a -> Type TyName uni ()
-    toTypeAst = toBuiltinTypeAst
-
-results in
-
-    Expected kind '* -> *', but 'uni' has kind 'k -> *'
-
-see https://gitlab.haskell.org/ghc/ghc/-/issues/15710
```
We no longer have this issue, since `Contains` (previously named `Includes`) is kind-polymorphic and so we don't need to add a type equality constraint to the default implementation of `toTypeAst` and it just works.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Universe/Core.hs:
```diff
@@ -118,14 +136,44 @@ class Closed uni where
+instance (forall a b. (constr a, constr b) => constr (f a b)) => constr `Permits2` f
+
+type Permits :: (GHC.Type -> Constraint) -> k -> Constraint
+type family Permits constr
```
Tried, it didn't work, wrote a small comment about it.

effectfully commented:
Ok, this monster is ready for review (there's one more test that I want to add, but otherwise the PR seems fine to me). Make sure to make a really big cup of tea.

bezirg commented on plutus-core/plutus-core/src/PlutusCore/MkPlc.hs:
```diff
@@ -69,13 +72,25 @@ class TermLike term tyname name uni fun | term -> tyname name uni fun where
     typeLet = mkImmediateTyAbs
 
--- | Embed a type from a universe into a PLC type.
+-- TODO: make it @forall {k}@ once we have that.
```
What do these two TODOs 'mean'?

effectfully replied:
[This](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0099-explicit-specificity.rst).

I'll expand the comment.

kwxm commented on plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs:
```diff
@@ -201,5 +202,4 @@ instance ExMemoryUsage Char where
 
-instance ExMemoryUsage String where
-  memoryUsage string = ExMemory $ fromIntegral $ (sum $ fmap sizeOf string) `div` 8
+deriving via GenericExMemoryUsage [a] instance ExMemoryUsage a => ExMemoryUsage [a]
```
This worries me.  If I understand correctly (and I may well have got it completely wrong), this will get the memory usage of a built-in list by traversing the entire list and adding up the memory usage of its members, along with some contribution from the constructors.  I think there are two things wrong with this. (A) if we're applying a built-in function to a list then we'll have to get the memory usage of the list just before applying the function, and this will preform a potentially expensive traversal.  (B) I think it's the wrong metric: we probably want the length of the list here. If we want to eg append one list to another then the cost will depend on the length of the lists (in fact, just the first list) and we don't care whether the entries are small booleans or massive bytestrings, but the sizes of those will be included in the generic memory usage. 

See also the main comment on this PR.

effectfully replied:
That's a good point, but we were using `fmap` before this PR as well, so the PR doesn't change much, so let's ignore the problem for now.

kwxm reviewed (approved):
As far as I understand it, this looks fine. A number of issues relating to costing occurred to me though.

Should we make lists (and other complicated buili-in types) carry their sizes along with them?  The size of a constant has to supplied to a costing function every time the list is used as an argument to a builtin, so it would be good if we could retrieve it with a constant-time operation, not, for example, by calling `length`,  which would itself increase the execution time of the program by some variable amount.  This would mean that functions returning lists etc would have to keep track of the size and bundle it together with the result, and this itself would have a cost, but it might be a smaller cost and it's one which would only be incurred once (we do calculate the memory consumption of our current buittin constants every time they're used,  but for most of them that's quite a fast calculation). Maybe I'm wrong: it could be that calculating the size of a list is so fast that we don't care, and in fact maybe the budgeting benchmarks would include that cost anyway. I'll think about that a bit more.

What would happen with, say, trees?  Some operations might depend on depth, others on total numbers of entries (and you surely wouldn't want to recalculate either of those every time you want to insert a new element into a large tree), so is there even a well-defined notion of size for such things?

Secondly, what type of functions are we going to allow on built-in types?  This PR provides lists and pairs, but doesn't provide any operations on them: presumably that'll happen in a later PR.  Things like `length` and `append` for lists should be OK (modulo how we obtain the sizes of lists : see the comments in `ExMemory.hs`), but I think that higher-order things like `map`, `fold`, and `filter` could be problematic.

For example, if we looking at `map f [x1, ..., xn]` then the time taken is going to be approximately the sum of the times taken to execute `f x1`, `f x2`,... , `f xn` and I don't see how we can possibly calculate this efficiently.  Consider `map (addInteger 1234567) [1,2, 89237894278937489, 4385908239085832482930489023894, ...]`.  To calculate in advance the cost of evaluating this we'd have to traverse the list, get the memory usage `mₖ` of each member, run the costing function for `addInteger` on `(1234567, mₖ)`, and then add up the results for each `k` (plus some overhead for traversal). This could be very expensive itself, and it's not clear to me how we'd know that this is the calculation we want to perform. The `map` builtin could provide a costing function, but it'd have to be a _higher-order_ costing function that would take the costing function for f as an argument.

So I think that if we want to keep control of costing, the functions we provide for operating on built-in polymorphic types will probably have to be constrained by how easy they are to cost.  That could be an unpleasant trade-off: a built-in `fold` could be much faster than a PLC implementation, but much more difficult to assign a cost to (even if it is a hundred times faster).  The difference is that with a PLC version we calculate the costs as we go along, but with a Haskell version (at least as things are at the moment) we have to calculate a single cost in advance.  I suppose we could have builtins which take account of budget constraints during their execution, but that could be unpleasant.

effectfully commented on plutus-core/untyped-plutus-core/test/Evaluation/Golden/iteAtIntegerFullyApplied.type.golden:
```diff
@@ -1,6 +1,6 @@
   '(con string "11 <= 22")'.
 Expected type
-  '(con string)',
+  '[(con []) (con char)]',
```
`String` is literally `[Char]` and due to `Note [Normalization of built-in types]` we have to normalize polymorphic built-in types by replacing intra-universe application with `TyApp`, which is how `(con string)` becomes `[(con []) (con char)]`.

kwxm replied:
Right, but there are a number of golden tests which are supposed to fail but I think they're now failing for the wrong reason.  This one is instantiating `ifThenElse` at `integer` but returning a string, but now the failure's because it's got the wrong sort of string, I think.

effectfully replied:
The diff is

```
- '(con string)',
+ '[(con []) (con char)]',
```

and the rest 

```
(Left Type mismatch at () in term
  '(con string "11 <= 22")'.
Expected type
  <...>,
found type
  '(con integer)')
```

is still the same. I.e. the expected type is now reported in normalized form (the new one) and that's it. Or am I missing something?

effectfully replied:
(I think the diff is fine and Kenneth is vacationing, so I'm resolving this)

effectfully commented on nix/pkgs/haskell/materialized-unix/.plan.nix/Win32-network.nix:
```diff
@@ -108,4 +108,4 @@
       };
-    } // rec { src = (pkgs.lib).mkDefault .././.source-repository-packages/18; }
\ No newline at end of file
+    } // rec { src = (pkgs.lib).mkDefault .././.source-repository-packages/19; }
```
I've added a new pinned dependency (`flat`) on a GitHub repo (rather than hackage). I think this is why `.nix` files needed to be updated. @michaelpj is everything fine here?

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs:
```diff
@@ -0,0 +1,233 @@
+    show DefaultUniUnit                = "unit"
+    show DefaultUniBool                = "bool"
+    show DefaultUniProtoList           = "[]"
+    show DefaultUniProtoTuple          = "(,)"
```
Indeed, I very much agree, especially given that we already have `unit` and not `()`. I'll fix it.

effectfully commented on plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs:
```diff
@@ -129,6 +141,7 @@ instance KnownType term Void where
 type instance ToBinds Void = '[]
 
+-- See Note [Representable built-in functions over polymorphic built-in types].
 -- Every polymorphic function ignores the memory annotation of its argument. This is due to the fact
```
Yep, thanks, I'll fix it.

effectfully commented:
> Secondly, what type of functions are we going to allow on built-in types? This PR provides lists and pairs, but doesn't provide any operations on them: presumably that'll happen in a later PR.

The PR provides `null`, `head` and `tail` (in `PlutusCore.Examples.Builtins`), which together give you `caseList`, in terms of which you can define folds or anything (see `Evaluation.DynamicBuiltins.Definition`). I.e. you can't construct lists **inside** Plutus Core right now, but you can compile them from Plutus Tx and deconstruct them with pattern matching.

> So I think that if we want to keep control of costing, the functions we provide for operating on built-in polymorphic types will probably have to be constrained by how easy they are to cost. 

I believe this is exactly right.

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+type ValueOf :: (Type -> Type) -> Type -> Type
+data ValueOf uni a = ValueOf !(uni (T a)) !a
+
+-- | A class for enumerating types and fully instantiated type formers that @uni@ contains.
```
multiline haddock

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+-- per type from the universe and you'll get 'Includes' for free.
+type Contains :: forall k. (Type -> Type) -> k -> Constraint
+class uni `Contains` a where
+    knownUni :: uni (T a)
```
This is an odd name, maybe? If we had `type Tag uni a = uni (T a)` then maybe that suggests this should be `knownTag`? or just `tag`? Or some other word for tag (`UniTag`?)

effectfully replied:
So I use `uni` everywhere to mean either "a universe" or "a type from the universe" depending on whether we're at the type or the term level. Analogously to `fun` meaning either "a set of built-in functions" or "a built-in function from the set of built-in functions" depending on whether we're at the type or the term level. I could make it `tag`, but I like that `uni` is more specific, even if it can mean two different things depending on where we are (which I think is pretty much never a problem just like calling a `newtype` an its constructor the same way is pretty much never a problem). Anyway, we could do that, but then it's going to touch literally everything related to universes and I'd prefer to do that in a separate PR (or something else can do it once this PR is merged).

effectfully replied:
I.e. I treat "uni" at the term level as a shorthand for "type tag".

michaelpj replied:
Okay, well _I_ find that punning confusing. I think I've been confused about this in the past too. I also find the distinction between types in a universe and tags in a universe sometimes confusing. Maybe this is no problem when you're used to it, I don't know if it's worth optimizing legibility for people who would be confused by this.

In the case of the tags, my confusion basically boils down to seeing `x :: uni (T a)` and going "x is a universe? what?" and then remembering that it doesn't work like that. Basically I learned to read `uni (T a)` as "tag". But then why not just help me and write `Tag`?

effectfully replied:
> I also find the distinction between types in a universe and tags in a universe sometimes confusing.

Semantics versus syntax pretty much. As the docs say:

> A universe is a collection of tags for types. It can be finite like
>
>             data U a where   
>                 UUnit :: U ()      
>                 UInt  :: U Int
> 
> (where 'UUnit' is a tag for '()' and 'UInt' is a tag for 'Int') or <...>
>
> We say that a type is in a universe whenever there is a tag for that type in the universe.

so I think this particular distinction is well-specified. But maybe now I need to clarify that a tag is not necessary a single constructor and can be an application of a constructor to another constructor/application.

> In the case of the tags, my confusion basically boils down to seeing x :: uni (T a) and going "x is a universe? what?"

We'd had `x :: uni a` up until this PR and it always meant "a type tag for `a` in `uni`". `x` never was a universe.

> Basically I learned to read `uni (T a)` as "tag". But then why not just help me and write `Tag`?

Right, that does seem better, I'll try it out.

michaelpj replied:
> We'd had x :: uni a up until this PR and it always meant "a type tag for a in uni". x never was a universe.

Yeah, and I think I was at least somewhat confused every time I saw it :D

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+knownUniOf :: uni `Contains` a => proxy a -> uni (T a)
+knownUniOf _ = knownUni
+
+-- | Wrap a value into @Some (ValueOf uni)@, given its explicit type tag.
```
more evidence that we should name `uni (T a)`

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+knownUniOf _ = knownUni
+
+-- | Wrap a value into @Some (ValueOf uni)@, given its explicit type tag.
+someValueOf :: forall a uni. uni (T a) -> a -> Some (ValueOf uni)
```
`someTaggedValue`?

michaelpj replied:
(I find the "tag" naming very helpful, hence why I'm suggesting it all over the place)

effectfully replied:
Even if we end up having tags in some form, I'd like to keep it `someValueOf`, 'cause it receives a type tag and a value **of** that type and literally returns a `Some (ValueOf uni)`.

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+someValue :: forall a uni. uni `Includes` a => a -> Some (ValueOf uni)
+someValue = someValueOf knownUni
+
+-- | A monad to decode types from a universe in.
```
Maybe explain why those monad?

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+
+    -- | Encode a type as a sequence of 'Int' tags.
+    -- The opposite of 'decodeUni'.
+    encodeUni :: uni a -> [Int]
```
This encodes a type in a universe, not the universe itself, so maybe it should be called something else? In fact, isn't this encoding the _tag_?

michaelpj replied:
Also, why *is* encoding/decoding in `Closed`? The description of `Closed` is all about `Everywhere`...

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+    i:is <- get
+    i <$ put is
+
+-- It's not possible to return a @forall@ from a type family, let alone compute a proper
```
too oblique!

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+class    (forall a b c. (constr a, constr b, constr c) => constr (f a b c)) => constr `Permits3` f
+instance (forall a b c. (constr a, constr b, constr c) => constr (f a b c)) => constr `Permits3` f
+
+-- I tried defining 'Permits' as a class but that didn't have the right inference properties
```
oh, here's the explanation

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+type Permits :: forall k. (Type -> Constraint) -> k -> Constraint
+type family Permits
+
+-- Implicit pattern matching on the kind.
```
:scream:

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+-- (i.e. I was getting errors in existing code). That probably requires bidirectional instances
+-- to work, but who cares given that the type family version works alright and can even be
+-- partially applied (the kind has to be provided immediately though, but that's fine).
+-- | @constr `Permits` f@ elaborates to one of
```
multiline haddock

michaelpj commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+-- Required only by 'Permits0' for some reason.
+{-# LANGUAGE UndecidableSuperClasses  #-}
+
+module Universe.Core
```
Maybe we should split this a little? Like the encoding/decoding bit could be separate, except it's tied into `Closed` and I'm not sure why.

effectfully replied:
So what makes a universe closed is that all of its type formers are fixed and no new ones can be added. E.g. quoting [this](https://golem.ph.utexas.edu/category/2011/04/homotopy_type_theory_v.html) (cut a little):

> the distinction as between an ‘open’ universe (to which new formers could be added) and a ‘closed’ universe (for which all the formers have been listed)

`Everywhere` is one way to enforce closedness and having a function that attempts to parse any sequence of integer tags as an encoded type tag is another, because that function is not extensible and so has to cover all possible cases and that is only possible in a closed universe. Plus, if you look at

```haskell
instance Closed uni => Hashable (SomeTypeIn uni) where
    hashWithSalt salt (SomeTypeIn uni) = hashWithSalt salt $ encodeUni uni

instance (Closed uni, uni `Everywhere` Hashable) => Hashable (ValueOf uni a) where
    hashWithSalt salt (ValueOf uni x) =
        bring (Proxy @Hashable) uni $ hashWithSalt salt (SomeTypeIn uni, x)
```

you can see that the constraints follow the same pattern as with `Eq`, `Show` etc, but we encode type tags to hash them in the `SomeTypeIn` instance. I.e. that notion of type tags being expressible as sequences of integer tags seems very interconnected with the notion of closedness under any constraints.

I could separate encoding/decoding into a separate class, but I don't see what that would gain us, apart from the opportunity to split up the module, half/most of which is literally documentation anyway. If anything, I'd separate instances and their docs (which amounts to a bit fewer than 200 lines). In some commits I did separate things into two modules, but then I decided that 750 lines is not a big deal, given that half/most of it is comments and merged them back.

michaelpj replied:
I get that it's not *so* much code, but it's quite a lot of conceptual weight, partly because of the docs! And there's all sorts of bits and pieces mixed in. I don't have a really good suggestion, though.

> Everywhere is one way to enforce closedness and having a function that attempts to parse any sequence of integer tags as an encoded type tag is another, because that function is not extensible and so has to cover all possible cases and that is only possible in a closed universe.

That seems to imply `Encodable => Closed`, but I'm not sure about the other direction. I don't *think* you can implement the encoding stuff with just `Everywhere`. It seems like there is maybe a refinement of something like `Constrainable = (has Everywhere) <= Partitionable (= tags can be disjointly assigned)`. Anyway, it's fine if this is more of a hassle than it's worth, I'm just trying to get my head around it.

effectfully replied:
I'll try it out.

effectfully replied:
> That seems to imply `Encodable => Closed`, but I'm not sure about the other direction.

I'm taking about semantics. If there's `Everywhere`, the universe is closed, using this definition of "closed":

> the distinction as between an ‘open’ universe (to which new formers could be added) and a ‘closed’ universe (for which all the formers have been listed)

If every type can be encoded/decoded, the universe is also closed in the same sense. It's not that `Closed` implies `Encodable` or something, it's just simply that open universes don't have these properties (how do you cover all the types in a certain instance/function at once if you don't know upfront what all these types are?).

I've expanded the comment a little to

> -- | A universe is 'Closed', if it's known how to constrain every type from the universe and
> -- every type can be encoded to / decoded from a sequence of integer tags.

So I really don't see much point in introducing a separate type class (and I can't come up with a good name for it anyway, it should be something universe-specific like `Closed` and related to encoding/decoding).

michaelpj commented on plutus-core/plutus-core/test/Evaluation/DynamicBuiltins/Common.hs:
```diff
@@ -21,7 +21,9 @@ import           Control.Monad.Except
 -- | Type check and evaluate a term that can contain dynamic built-ins.
 typecheckAnd
-    :: (MonadError (Error uni fun ()) m, ToBuiltinMeaning uni fun, GShow uni, GEq uni)
+    :: ( MonadError (Error uni fun ()) m, ToKind uni, HasUniApply uni, ToBuiltinMeaning uni fun
```
Yeah, some judicious constraint synonyms might be quite useful.

effectfully replied:
> A bit of grepping suggests that apart from in the typechecking code, `ToKind` almost always appears along with `HasUniApply` and usually also with `ToBuiltinMeaning`

We need `ToKind` and `HasUniApply` for type checking, but `HasUniApply` is also used for evaluation (see the `KnownType term (SomeConstantOf uni f reps)` instance in `Typed.hs`) without `ToKind` and `ToBuiltinMeaning` is used for type checking builtins (but not in the main type checking procedure) and evaluation. So we only use all the three in functions that type check entire programs. I guess I can define

```haskell
type Typecheckable uni fun = ToKind uni, HasUniApply uni, ToBuiltinMeaning uni fun
```

for that purpose.

michaelpj reviewed (commented):
Possibly more to come, but here's a batch.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:
```diff
@@ -91,14 +98,39 @@ newtype Normalized a = Normalized
+-- no big deal.
+-- | For getting the Plutus kind of a type from the universe.
+class ToKind (uni :: GHC.Type -> GHC.Type) where
+    toKind :: forall k (a :: k). uni (T a) -> Kind ()
```
Sure, do you have a suggestion?

> I was initially convinced `T` must be a type variable since that's always what one-letter types are. 

But it starts with a capital letter!

michaelpj replied:
Sure, so the "it's a type variable" and "it's an actual type" intuitions warred in my head and I was confused.

I note that if we did define `Tag`, then most uses of `T` would go away, and it would matter a *lot* less what we called it. Then we could give it a longer name and not care.

effectfully replied:
> I was initially convinced `T` must be a type variable since that's always what one-letter types are. Can we pick a different name?

Hm, I used that "escape" term and I liked it. What do you think about `Esc`? For example:

```haskell
data DefaultUni a where
    DefaultUniBool :: DefaultUni (Esc Bool)
    DefaultUniList :: DefaultUni (Esc [])    
```

That makes it more clear that `Esc` is just syntactic noise like `\`.

effectfully replied:
^ also pinging @kwxm

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Constant/Typed.hs:
```diff
@@ -396,6 +484,110 @@ instance KnownType term a => KnownType term (Emitter a) where
+    makeKnown = pure . fromConstant . unSomeConstant
+    readKnown = fmap SomeConstant . unliftSomeValue
+
+-- | 'SomeConstantOf' is similar to 'SomeConstant': while the latter is for unlifting any
```
Yeah, but I like `--` more. I don't mind changing that to multiline comments though.

michaelpj replied:
I'd prefer it. I find long comments with `--` harder to read and *much* harder to edit.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Normalize/Internal.hs:
```diff
@@ -118,12 +121,44 @@ was previously added to an environment (while handling the function application
+-- See Note [Normalization of built-in types].
+-- | Normalize a built-in type by replacing each application inside the universe with regular
+-- type application.
+normalizeUni :: forall k (a :: k) uni tyname. HasUniApply uni => uni (T a) -> Type tyname uni ()
```
> This feels a little odd to me, since in some ways this seems like the opposite of normalization, if normalization is reduction-like.

Normalization doesn't have to be reduction-like, it's literally a process of reducing something to a normal form, whatever that form is.

> Shouldn't regular type application of builtins "reduce" to application in the built-in universe, if anything?

That would be harder to implement, because collapsing a general `Type` into a `SomeTypeIn` is harder than expanding a `SomeTypeIn` into a `Type`. And it would also be impossible to do in the general case, 'cause you can't collapse an application of a built-in type to, say, a type variable into an intra-universe application as there are no type variables there. So I guess we could "reduce" application of built-in types in some cases and not reduce them in others and still make the whole thing work, but that requires substantially more logic and is also a lot harder to get right.

I'll expand the docs.

effectfully commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+-}
+
+-- See Note [Representing polymorphism].
+type T :: forall k. k -> Type
```
It's literally "any type whatsoever", hence the super general name (given that `Any` is already occupied).

effectfully commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+
+-- See Note [Representing polymorphism].
+type T :: forall k. k -> Type
+data T a
```
Will do.

effectfully commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+
+-- | Existential quantification as a data type.
+type Some :: forall a. (a -> Type) -> Type
+data Some f = forall x. Some !(f x)
```
I specifically abstracted the whole thing out of Plutus as it's a very general machinery, so this is kinda like a library, so it might eventually be studied by people and explicit bangs help readability a lot compared to a sneaky pragma. So I'm not against `StrictData` in general and we could have it in this file as well, but I'd prefer to keep the explicit bangs.

effectfully commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+type SomeTypeIn :: (Type -> Type) -> Type
+data SomeTypeIn uni = forall k (a :: k). SomeTypeIn !(uni (T a))
+
+data Kinded uni ta where
```
It's a GADT, `ta` is reified as `T (a :: k)`

michaelpj replied:
Ah right

effectfully commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+
+-- | A particular type from a universe.
+type SomeTypeIn :: (Type -> Type) -> Type
+data SomeTypeIn uni = forall k (a :: k). SomeTypeIn !(uni (T a))
```
I tried doing that (except with a `newtype` rather than `type`, 'cause I rarely like the idea of exporting a `type`) and I didn't like how it looked or the ergonomics or something. That `Tag` type then has to be of kind `forall k. (Type -> Type) -> k -> Type` and that kind-polymorphism is somewhat unfortunate... Anyway, I don't mind trying again, but I'll use `newtype` then.

michaelpj replied:
The idea with `type` is just that it's there for readability rather than anything else, but you could do a newtype also.

effectfully replied:
> you could do a newtype also.

That turned out to be a terrible idea. Reasons:

1. if we adopt using `Tag uni a` in place of `uni (T a)`, we'll end up having nested tagness like here:

```haskell
instance Parsable (SomeTypeIn (Kinded DefaultUni)) where
    parse "bool"       = Just . SomeTypeIn . Tag . Kinded $ Tag DefaultUniBool
```

(it took me several minutes to realize that in a more complex case)

2. we'd have to either duplicate a lot of the API or define type class methods over `Tag uni a` as well, which is somewhat annoying. For example:

```haskell
instance Closed DefaultUni where
    <...>
    bring p (Tag (DefaultUniProtoPair `DefaultUniApply` uniA `DefaultUniApply` uniB)) r =
        bring p (Tag uniA) $ bring p (Tag uniB) r
```

3. the user has to wrap/unwrap `Tag`. For example:

```haskell
instance Closed uni => Flat (SomeTypeIn uni) where
    <...>
    decode = decodeKindedUniFlat <&> \(SomeTypeIn (Tag (Kinded uni))) -> SomeTypeIn uni
```

so that `T` still pops up in user code.

4. but the real problem is that `Tag` is not a sound abstraction: we want to be able to call things like `geq` over it without unpacking it first, but we just can't: `Tag uni a` in its full looks like `Tag @k uni a` where `k` is the kind of `a`, and so if we define `GEq (Tag uni)` that's in fact `GEq (Tag @k uni)`, i.e. it hardcodes a specific kind, while with `GEq uni` we retain the full generality with `T` "escaping" types of arbitrary kinds. Using `Tag` does break our code and if it takes me some effort to deal with that broken abstraction, it's going to be much harder for someone who did not write that code.

I wouldn't mind making `Tag` a type synonym if you wish so, but I consider it pointless. Type synonyms are good when they don't leave their module or when they shorten a definition by a lot or when they are needed to simulate subtyping like in `lens`, but in our case it would be just two different equally short ways to write the same thing with the "undesired" one popping up in error messages etc, so that the user will have to learn what it means anyway.

michaelpj replied:
> Type synonyms are good when they don't leave their module or when they shorten a definition by a lot or when they are needed to simulate subtyping like in lens, but in our case it would be just two different equally short ways to write the same thing with the "undesired" one popping up in error messages etc, so that the user will have to learn what it means anyway.

I'm not sure I totally agree with that. I spend a lot of time reading type signatures, and a synonym can be quite helpful there. But I don't mind that much.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Constant/Typed.hs:
```diff
@@ -396,6 +484,110 @@ instance KnownType term a => KnownType term (Emitter a) where
+-- So be careful in the denotation of a builtin to unlift its arguments to what you promised to
+-- unlift them to in its type signature.
+type SomeConstantOf :: forall k. (GHC.Type -> GHC.Type) -> k -> [GHC.Type] -> GHC.Type
+data SomeConstantOf uni (f :: k) reps where
```
Nope, a leftover, will fix.

effectfully commented on plutus-core/plutus-core/src/Universe/Core.hs:
```diff
@@ -0,0 +1,756 @@
+
+-- | A value of a particular type from a universe.
+type ValueOf :: (Type -> Type) -> Type -> Type
+data ValueOf uni a = ValueOf !(uni (T a)) !a
```
(discussing elsewhere, so resolving)

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs:
```diff
@@ -164,7 +163,7 @@ deriving newtype instance ExMemoryUsage ExMemory
 
 -- See https://github.com/input-output-hk/plutus/issues/1861
-instance ExMemoryUsage (Some (TypeIn uni)) where
+instance ExMemoryUsage (SomeTypeIn uni) where
```
> The only thing that we care about are the sizes of elements of buitlin types, so I think we should be able to get rid of a lot of this.

The thing is, we probably do care about the size of the type tag as we do store it at runtime. This will be especially important once we have built-in lists and tuples as the type tag for, say, a list is a sequence of constructors of `DefaultUni`, not just a single one.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs:
```diff
@@ -0,0 +1,233 @@
+-- | The universe used by default and its instances.
```
I've refactored that. Now we have a separate `Universe` library (currently living under `plutus-core/plutus-core/src`, but we could move it elsewhere) consisting of an internal module (called `Core`) and a user-facing one just reexporting stuff from the internal module (we may end up having a richer structure of internal modules) and a single module in Plutus Core defining the default universe: `plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs`.

Seems fine?

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs:
```diff
@@ -0,0 +1,233 @@
+See Note [Stable encoding of PLC]
+-}
+
+instance Closed DefaultUni where
```
> or is there any way to extend it without copying everything?

But you don't need to copy anything? If you add a new built-in type, you'll just need to add one more case to functions `encodeUni`, `withDecodedUni` and `bring` and one more line to `Everywhere`. `Show` and `Parsable` are a bit more nasty as you may need to handle each partial application in its own way there like we do now, but extending the `Closed` instance is very straightforward and requires little typing.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Parsable.hs:
```diff
@@ -21,25 +23,97 @@ import           Data.ByteString (ByteString)
+lexical analysis upfront.
+
+So at the moment we don't do anything correctly. Neither PLC nor PIR can handle lists of lists
+and none of them can handle tuples at all ('Read' provides us with overloaded parsing for lists
```
I'm going to pretend that we don't have any problems with the parser within the scope of this PR.

> It would be a lot easier to use some different syntax like `<(con integer 5), (con bool True)>`

(note that it would be `(con (pair integer bool) <5, True>)` then. We can't put arbitrary ASTs inside polymorphic built-in types and `con` creates an "arbitrary AST"). Banning parentheses and square brackets inside the syntax of constants would certainly simplify parsing and I also think that might be a good idea that would probably allow us note to poke a hole the parser and instead parse constants in chunks like we do right now.

effectfully commented:
I think I've addressed all comments except refactoring `uni (Esc a)` into `Tag uni a` and then refactoring every single use of `uni` at the term level into quite less specific `tag`, which I'm not very fond of. @michaelpj if you'd still like me to do that, I suggest to do that in a subsequence PR and merge this one the moment CI is green.

michaelpj commented:
Eh, I don't mind _that_ much. Let's just leave it for now.

effectfully commented:
:tada:
